src/fmu/pem/pem_functions/effective_pressure.py

```diff
@@ -13,81 +16,171 @@
+    actual_fipnum_values = list(np.unique(fipnum_data[~fipnum_mask]).astype(int))
+
+    # Calculate bulk density for all time steps (needed for overburden pressure
+    # calculation)
+    # Bulk density calculation needs to be zone-aware for patchy cement models
+    fl_density = [fluid.density for fluid in fluid_props]
+
+    # Check if any zone uses patchy cement model and needs special handling
+    bulk_density_list = []
+    for time_idx, fl_dens in enumerate(fl_density):
+        bulk_dens_grid = np.ma.masked_array(
+            np.zeros(sim_init.poro.shape, dtype=float), mask=fipnum_mask
+        )
+
+        # Calculate bulk density per zone based on zone-specific model
+        for zone_idx, zone_region in enumerate(rock_matrix.zone_regions):
+            # Get all FIPNUM values for this zone using input_num_string_to_list
+            fipnum_values = input_num_string_to_list(
+                zone_region.fipnum, actual_fipnum_values
+            )
+
+            # Build combined mask for all FIPNUMs in this zone using vectorized
+            # operation
+            zone_mask = np.isin(fipnum_data, fipnum_values) & ~fipnum_mask
+
+            # Check if this zone uses patchy cement model
+            is_patchy_cement = zone_region.model.model_name == RPMType.PATCHY_CEMENT
+
+            if is_patchy_cement:
+                # Get cement fraction and density for patchy cement model
+                cement_fraction = zone_region.model.parameters.cement_fraction
+                cement_density = rock_matrix.minerals[rock_matrix.cement].density
+            else:
+                cement_fraction = None
+                cement_density = None
+
+            # Calculate bulk density for this zone
+            zone_bulk_density = estimate_bulk_density(
+                porosity=np.ma.masked_where(~zone_mask, sim_init.poro),
+                fluid_density=[np.ma.masked_where(~zone_mask, fl_dens)],
+                mineral_density=np.ma.masked_where(~zone_mask, matrix_props.density),
+                patchy_cement=is_patchy_cement,
+                cement_fraction=cement_fraction,
+                cement_density=cement_density,
             )
-            for (sim_date, dens) in zip(sim_rst, bulk_density)
-        ]
-    else:  # overburden_pressure.type == 'trend':
-        eff_pres = [
-            estimate_effective_pressure(
-                formation_pressure=sim_date.pressure * 1.0e5,
-                bulk_density=dens,  # type: ignore
-                reference_overburden_pressure=overburden_pressure_from_trend(
-                    inter=overburden_pressure.intercept,
-                    grad=overburden_pressure.gradient,
-                    depth=sim_init.depth,
-                ),
+
+            # Merge zone bulk density into full grid
+            bulk_dens_grid[zone_mask] = zone_bulk_density[0][zone_mask]
+
+        bulk_density_list.append(bulk_dens_grid)
+
+    bulk_density = bulk_density_list
```

Comment:
I think you can simplify:
- why just use bulk_density_list later also
- or rename bulk_density_list to bulk_density initially